text/3374-cargo-target-features.md

```diff
@@ -0,0 +1,177 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This adds additional complexity to `Cargo.toml` potentially making it more difficult to understand.
+* Users may be easily confused between the difference of `enable-features` and `required-features`.
+  Hopefully clear and explicit documentation may help avoid some of that confusion.
+  The error messages with `required-features` may also be extended to mention `enable-features` as an alternative (as well as other situations like `cargo install` failing to find a binary).
+* It may not be clear that features are unified across all targets.
+  This may particularly come into play where it may not be clear that an optional dependency suddenly becomes *available* to all the other targets when the `enable-features` causes it to be included.
+  A similar situation arises with dev-dependencies, where a user may get confused when referencing a dev-dependency outside of a `#[cfg(test)]` block or module, which causes an error.
+* This may not have the same clarity as explicit dependencies as proposed in [RFC 2887](https://github.com/rust-lang/rfcs/pull/2887).
+* There may be increased confusion and complexity regarding the interaction of various cargo commands and this feature (such as `cargo tree` mentioned above which has no concept of cargo target selection).
+* There may be confusion about the interaction with `--no-default-features`.
+  `--no-default-features` will continue to only affect the `default` feature.
+  Features enabled by `enable-features` will be enabled even with `--no-default-features`.
+  This may be confusing and should be mentioned in the documentation.
+  * There are some alternatives to avoiding a target from being built, such as not including it in the CLI options, using `required-features` instead, or disabling it with fields such as `test=false`.
+    However, not all of these options may be satisfying.
+* This may add significant complexity to Cargo's implementation.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Change the `required-features` behavior
+
+* `required-features` could be changed to behave the same as `enable-features` described in this RFC (possibly over an Edition).
+However, as outlined in the [Relationship with `required-features`](#relationship-with-required-features) section, there are some use cases where the present behavior of `required-features` is desirable.
+This could also lead to a breaking change for some projects if it started building targets that were previously not included.
+* Instead of adding a separate field that lists features, a `force-enable-features = true` field could be added to change the behavior of `required-features` to have the behavior explained in this RFC.
+  That might be less confusing, but would prevent the ability to have both behaviors at the same time.
```

Comment:
Can you speak to when it is desirable to have both `enable-features` _and_ `required-features` at the same time? (I'm assuming `force-enable-features` could be set per-target).

Reply:
I would expect it to be rare. A possible example would be:

```toml
[[example]]
name = "db"
enable-features = ["dep:clap"]
required-features = ["sql"]
```

Here the example would only be built if `-F sql` were specified. This could be done if you don't want a heavy dependency (like a database dependency) to always be built when running `cargo test`. The presence of `clap` isn't the primary focus of the example, so it is a hidden aspect of the example, and will get enabled automatically if the example is built.

I think a con for `force-enable-features = true` I forgot to list is that it would require entering two lines to get the behavior that I think most users will want.

Reply:
I quite like the distinction between "enable this target if X features are enabled" _in addition to_ "if this target is enabled, build it with Y features". If I'm understanding correctly, `force-enable-features` turns that into an either/or, which I think is more work for less utility overall. :-1: on `force-enable-features` if that's the case.